# Algoritmos de optimización - Seminario<br>
Nombre y Apellidos:   <br>
Url: https://github.com/.../03MAIR---Algoritmos-de-Optimizacion---2019/tree/master/SEMINARIO<br>
Problema:
> 1. Sesiones de doblaje <br>
>2. Organizar los horarios de partidos de La Liga<br>
>3. Combinar cifras y operaciones

Descripción del problema:

#### **Problema 3.** 
**Combinar cifras y operaciones**

- El problema consiste en analizar el siguiente problema y diseñar un algoritmo que lo resuelva.

- Disponemos de las 9 cifras del 1 al 9 (excluimos el cero) y de los 4 signos basicos de las operaciones fundamentales: suma(+), resta(-), multiplicacion(*), y division(/)

- Debemos combinarlos alternativamente sin repetir ninguno de ellos para obtener una cantidad dada. Une ejemplo seria para obtener el 4:

4+6-6/3*1 = 4



- Debe analizarse el problema para encontrar todos los valores enteros posibles planteando las siguientes cuestiones:
¿Que valor maximo y minimo se puede obtener segun las condiciones del problema?
¿Es posible encontrar todos los valores enteros posibles entre dicho minimo y maximo?

Nota: es posible usar la funcion de python "eval" para evaluar una expresión:

expresion = "4-2+6/3*1"

print(eval(expresion))

4.0

(*) La respuesta es obligatoria





                                        

In [1]:
import itertools
import operator

###**Lista de numeros del 1 al 9**

In [2]:

nums= list(range(1,10,1))

nums

[1, 2, 3, 4, 5, 6, 7, 8, 9]

###De forma aleatoria e iterativa se realizan las **permutaciónes** en cinco cifras y calculando todas las posibles soluciones.



In [3]:
_oper= []
_resultados= []

*Ciclo iterativo*

In [4]:
for i in itertools.permutations(nums, 5):
  oper= str(i[0])+ '-' + str(i[1]) + '+' + str(i[2]) + '/' + str(i[3]) + '*' + str(i[4])
  #print(oper)
  _oper.append(oper)
  resultados = eval(oper)
  _resultados.append(resultados)

In [5]:
_resultados

[2.75,
 3.5,
 4.25,
 5.0,
 5.75,
 1.4,
 2.5999999999999996,
 3.2,
 3.8,
 4.3999999999999995,
 1.0,
 1.5,
 2.5,
 3.0,
 3.5,
 0.7142857142857142,
 1.1428571428571428,
 1.5714285714285712,
 2.4285714285714284,
 2.8571428571428568,
 0.5,
 0.875,
 1.25,
 1.625,
 2.375,
 0.33333333333333326,
 0.6666666666666665,
 1.0,
 1.333333333333333,
 1.6666666666666665,
 5.666666666666666,
 7.0,
 8.333333333333332,
 9.666666666666666,
 11.0,
 1.4000000000000004,
 3.8000000000000007,
 4.6000000000000005,
 5.4,
 6.2,
 1.0,
 2.333333333333333,
 3.666666666666666,
 4.333333333333333,
 5.0,
 0.7142857142857142,
 1.8571428571428568,
 2.4285714285714284,
 3.571428571428571,
 4.142857142857142,
 0.5,
 1.5,
 2.0,
 2.5,
 3.5,
 0.33333333333333326,
 1.2222222222222223,
 1.6666666666666665,
 2.1111111111111107,
 2.5555555555555554,
 5.666666666666667,
 9.0,
 10.666666666666668,
 12.333333333333334,
 14.0,
 2.75,
 6.5,
 7.75,
 9.0,
 10.25,
 1.5,
 2.3333333333333335,
 4.833333333333334,
 5.666666666666667,
 6.5,
 1.1

In [6]:
import numpy as np
Max = np.max(_resultados)
Min = np.min(_resultados)
print('Se calculan los valore máximos y minimos de los resultados')
print('El valor máximo es =' + str(Max), '\n' 'El valor Mínimos es = ' + str(Min))

Se calculan los valore máximos y minimos de los resultados
El valor máximo es =77.0 
El valor Mínimos es = -7.25


In [7]:
digits= 9
pos= 4
nums= pow(digits,pos)
nums

6561

In [8]:
len(_oper)

15120

###(*)¿Cuantas posibilidades hay sin tener en cuenta las restricciones?<br>


```
#código

digits= 9
pos= 4

nums= np.pow(digits,pos)
nums

```

**6561** posibilidades sin restricciones




###¿Cuantas posibilidades hay teniendo en cuenta todas las restricciones.
```
#código

 len(_oper)
```
**30242** posibilidades teniendo en cuenta restricciones.



Modelo para el espacio de soluciones<br>
(*) ¿Cual es la estructura de datos que mejor se adapta al problema? Argumentalo.(Es posible que hayas elegido una al principio y veas la necesidad de cambiar, arguentalo)


###Respuesta

Los reultados enlistados permiten permiten salvar por medio de condiciones de forma iterativa y sin la necesidad de hacer un sort ya que van de forma ordenada. 


###Según el modelo para el espacio de soluciones<br>
####(*)¿Cual es la función objetivo?
```
# código

i = numeros
_i= i[1],i[2],i[3],i[4]
```
$f(_i) = \frac{i1 -i2 + i3}{i4 * i5}$


####(*)¿Es un problema de maximización o minimización?

Aplica para ambos casos

####Diseña un algoritmo para resolver el problema por fuerza bruta

Respuesta

Este es un problema de detección de anagramas para cadenas de caracteres. por ejemplo (Fresa/Frase) son anagramas que mediante el reordenamiento de los caracteres y un diccionario puede detectar si son anagramas, lo cuál se puede resolver mediante fuerza bruta.

In [9]:
def anagram(cad1,cad2):
    c1 = [0]*26
    c2 = [0]*26

    for i in range(len(cad1)):
        pos = ord(cad1[i])-ord('a')
        c1[pos] = c1[pos] + 1

    for i in range(len(cad2)):
        pos = ord(cad2[i])-ord('a')
        c2[pos] = c2[pos] + 1

    j = 0
    OK = True
    while j<26 and OK:
        if c1[j]==c2[j]:
            j = j + 1
        else:
            OK = False

    return OK

print(anagram('covid','divoc'))
print(anagram('fuego', 'arena'))

True
False


####Calcula la complejidad del algoritmo por fuerza bruta

Respuesta

In [10]:
from time import time

tiempo_inicial = time() 
anagram('covid','divoc')
tiempo_f = time() 

ejecución = tiempo_f - tiempo_inicial
print('EL tiempo estimado de la ejecución es de = ' + str(ejecución) + ' segundos')

EL tiempo estimado de la ejecución es de = 6.4849853515625e-05 segundos


(*)Diseña un algoritmo que mejore la complejidad del algortimo por fuerza bruta. Argumenta porque crees que mejora el algoritmo por fuerza bruta

Respuesta

In [11]:
def Anagram(str1, str2): 

    return sorted (str1) == sorted (str2)

print(Anagram('colinas','nicolas'))

True


(*)Calcula la complejidad del algoritmo 

Respuesta

In [153]:
from time import time

tiempo_inicial = time() 
Anagram('covid','divoc')
tiempo_f = time() 

ejecución1 = tiempo_f - tiempo_inicial
print('EL tiempo estimado de la ejecución es de = ' + str(ejecución1) + ' segundos')

EL tiempo estimado de la ejecución es de = 5.245208740234375e-05 segundos


**R=**Comparando las funciones generadas para los algoritmos de ejecución del anagrama, se representa una mejor en 'Anagram' ya que el hecho de usar .sort, le permite no repetir la posición de ninguna de las letras en cualquier ordenamiento, sin embargo en 'anagram' no importa la cantidad de letras que tenga siempre y cuando se cumpla que se contengan las mismas letras en ambas cadenas. 

Añadiendo a esto, el tiempo de ejecución es menor debido a que el ciclo en 'anagram' es iterativo, mientras que en 'Anagram' hace el conteo de letras y las saca del bucle una vez que ya las ha utilizado sin cumplir todo el ciclo. 

Tambien sin mencionar que son menos lineas de código y la complejidad resuelve en O(n).

Desventajas: No reconoce como valido signos de puntuación distintos así como mayusculas son diferentes a minpusculas.

In [13]:
print('Comparamos el tiempo de anagram vs Anagram','\n' 'T(anagram)' + str(ejecución) + 'segundos', '\n' 'T(Anagram)' + str(ejecución1) + 'segundos <----') 

Comparamos el tiempo de anagram vs Anagram 
T(anagram)6.4849853515625e-05segundos 
T(Anagram)5.245208740234375e-05segundos <----


####Según el problema (y tenga sentido), diseña un juego de datos de entrada aleatorios


Respuesta

In [33]:
import random 
WORDS = ("fresa", "frase", "alegan", "angela", "colinas","nicolas","riesgo","sergio","mora","roma","cardiografía","radiografía","desamparador", "desparramado","materialismo","memorialista","escabullimiento","bulliciosamente") 

Aplica el algoritmo al juego de datos generado

Respuesta

In [147]:
word1 = random.choice(WORDS) 
word2 = random.choice(WORDS)

while word2 == word1:
 word2 = random.choice(WORDS)
print('Las palabras escogidas son:','\n' + str(word1),'\n'+ '&','\n'+ str(word2)) 

if Anagram(word1,word2) == True:
   print('Las palabras ' + str(word1) + ' y ' + str(word2) + ' son anagramas.')
else:
   print('Las palabras ' + str(word1) + ' y ' + str(word2) + ' NO son anagramas.', "\n" + 'Vuelva a intentar')

Las palabras escogidas son: 
colinas 
& 
nicolas
Las palabras colinas y nicolas son anagramas.


Enumera las referencias que has utilizado(si ha sido necesario) para llevar a cabo el trabajo
(*)1 . https://www.datacamp.com/community/tutorials/python-while-loop?utm_source=adwords_ppc&utm_campaignid=1455363063&utm_adgroupid=65083631748&utm_device=c&utm_keyword=&utm_matchtype=b&utm_network=g&utm_adpostion=&utm_creative=332602034364&utm_targetid=aud-392016246653:dsa-429603003980&utm_loc_interest_ms=&utm_loc_physical_ms=1010045&gclid=Cj0KCQiAwMP9BRCzARIsAPWTJ_FfygOSSf3QsYxok7C9_7ABDL2aIeGrtzBkUkpR8l93BCBGNFXmHxQaAshqEALw_wcB

(*)2 .https://docs.python.org/2/library/itertools.html


(*)3. 1st September 2017, 12:18 AM https://www.sololearn.com/Discuss/674616/how-to-check-if-two-string-are-anagram
David Ashton 

Respuesta

####Describe brevemente las lineas de como crees que es posible avanzar en el estudio del problema. Ten en cuenta incluso posibles variaciones del problema y/o variaciones al alza del tamaño

Respuesta

EL coste computacional del algoritmo ya es bajo en cuanto a su procesamiento por lo que debería mejorarse añadiendo o asignando dentro de un one hot encoding aquellas letras similares en minusculas y mayusculas así como aquellas con acento para que el valor de la letra sea el mismo asignado aunque el tamaño de la matriz incrementaría en gran parte el coste computacional.